dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,263 @@
+
+        if (WI_IsFlagSet(details.Kind(), PushNotificationRegistrationKind::PushTrigger))
+        {
+            bool taskRegistered = false;
+            for (auto task : BackgroundTaskRegistration::AllTasks())
+            {
+                if (task.Value().Name() == backgroundTaskName)
+                {
+                    taskRegistered = true;
+                    break;
+                }
+            }
+
+            if (!taskRegistered)
+            {
+                BackgroundTaskBuilder builder;
+                builder.Name(backgroundTaskName);
+
+                PushNotificationTrigger trigger{};
+                builder.SetTrigger(trigger);
+
+                if (taskClsid != winrt::guid())
+                {
+                    auto builder5 = builder.try_as<winrt::IBackgroundTaskBuilder5>();
+                    if (IsPackagedProcess() && builder5)
+                    {
+                        builder5.SetTaskEntryPointClsid(taskClsid);
+                        winrt::com_array<winrt::IBackgroundCondition> conditions = details.GetConditions();
+                        for (auto condition : conditions)
+                        {
+                            builder.AddCondition(condition);
+                        }
+                    }
+                }
+
+                registeredTask = builder.Register();
+            }
+        }
+
+        if (WI_IsFlagSet(details.Kind(), PushNotificationRegistrationKind::ComActivator))
+        {
+            if (taskClsid != winrt::guid())
+            {
+                {
+                    auto lock = g_lock.lock();
+                    // Define handle that will be set during background task execution
+                    g_waitHandleForArgs = wil::unique_handle(CreateEvent(nullptr, FALSE, FALSE, nullptr));
+                }
+
+                THROW_IF_FAILED(::CoRegisterClassObject(
+                    taskClsid,
+                    winrt::make<PushNotificationBackgroundTaskFactory>().get(),
+                    CLSCTX_LOCAL_SERVER,
+                    REGCLS_MULTIPLEUSE,
+                    &cookie));
+            }
+        }
+        return PushNotificationRegistrationToken{ cookie, registeredTask };
+    }
+
+    void PushNotificationManager::UnregisterActivator(PushNotificationRegistrationToken const& token, PushNotificationRegistrationKind const& kind)
+    {
+
+        if (WI_IsFlagSet(kind, PushNotificationRegistrationKind::PushTrigger))
+        {
+            for (auto task : BackgroundTaskRegistration::AllTasks())
+            {
+                if (task.Value().Name() == backgroundTaskName)
+                {
+                    task.Value().Unregister(false /*cancel task*/);
+                }
+            }
+        }
+
+        if (WI_IsFlagSet(kind, PushNotificationRegistrationKind::ComActivator) && (token.Cookie() != 0))
+        {
+            THROW_IF_FAILED(::CoRevokeClassObject(static_cast<DWORD>(token.Cookie())));
```

Comment:
Consider just `LOG_IF_FAILED` here - failure to unregister probably shouldn't be fatal.